src/test/gather_node_test.cpp

```diff
@@ -0,0 +1,189 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include <algorithm>
+#include <sstream>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "../dl_node.hpp"
+#include "../entry_node.hpp"
+#include "../exit_node.hpp"
+#include "../gathernodeinputhandler.hpp"
+#include "../modelconfig.hpp"
+#include "../nodeinputhandler.hpp"
+#include "../pipeline.hpp"
+#include "../pipeline_factory.hpp"
+#include "../pipelinedefinition.hpp"
+#define DEBUG
+#include <cstdio>
+
+#include <stdlib.h>
+
+#include "../dlnodesession.hpp"
+#include "../logging.hpp"
+#include "../modelinstance.hpp"
+#include "../prediction_service_utils.hpp"
+#include "../status.hpp"
+#include "test_utils.hpp"
+
+using namespace ovms;
+using namespace tensorflow;
+using namespace tensorflow::serving;
+
+using testing::_;
+using testing::ElementsAre;
+using testing::Return;
+
+class GatherNodeInputHandlerTest : public ::testing::Test {};
+
+TEST_F(GatherNodeInputHandlerTest, ThreePredecessorNodesWithSubsessionSize2) {
+    // simulate all 3 inputs comming from different predecessor nodes
+    // with session demultiplexed to 2 shards
+    const std::vector<std::string> inputNames{"a", "b", "c"};
+    const std::vector<size_t> shape{1, 10};
+    const InferenceEngine::Precision precision{InferenceEngine::Precision::FP32};
+    const InferenceEngine::Layout layout{InferenceEngine::Layout::NC};
+    std::vector<float> blobData{-1, 4, 5, 12, 3, 52, 12, 0.5, 9, 1.67};
+    const InferenceEngine::TensorDesc desc{precision, shape, layout};
+    InferenceEngine::Blob::Ptr inputBlob = InferenceEngine::make_shared_blob<float>(desc, blobData.data());
+    const uint32_t shardsCount = 2;  // subsessionSize/demultiplyCount
+    GatherNodeInputHandler gInputHandler(inputNames.size(), shardsCount);
+    for (session_id_t j = 0; j < shardsCount; ++j) {
+        for (size_t i = 0; i < 3; ++i) {
```

Comment:
Here instead of hardcoding `3` you can use `inputNames.size()`, this way we could modify this test in the future by just changing vector in line 56.